pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go

```diff
@@ -551,3 +555,78 @@ func TestTenantStreamingCheckpoint(t *testing.T) {
 	<-time.NewTimer(3 * time.Second).C
 	require.Equal(t, [][]string{{"2"}}, c.getDestTenantSQL().QueryStr(t, "SELECT * FROM d.t2"))
 }
+
+func TestTenantStreamingTimestampProtection(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	skip.UnderRaceWithIssue(t, 83867)
+	skip.UnderStressRace(t, "slow under stressrace")
+
+	rnd, _ := randutil.NewTestRand()
+	ctx := context.Background()
+
+	loadProducerPTSRecord := func(c *tenantStreamingClusters, producerJobID int) *ptpb.Record {
+		jobRegistry := c.srcSysServer.JobRegistry().(*jobs.Registry)
+		job, err := jobRegistry.LoadJob(context.Background(), jobspb.JobID(producerJobID))
+		require.NoError(t, err)
+
+		serverCfg := c.srcSysServer.DistSQLServer().(*distsql.ServerImpl).ServerConfig
+		ptp := serverCfg.ProtectedTimestampProvider
+
+		payload := job.Payload()
+
+		var record *ptpb.Record
+		require.NoError(t, serverCfg.DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) (err error) {
+			record, err = ptp.GetRecord(ctx, txn, payload.GetStreamReplication().ProtectedTimestampRecordID)
+			return err
+		}))
+
+		return record
+	}
+
+	loadIngestProgress := func(c *tenantStreamingClusters, ingestionJobID int) *jobspb.StreamIngestionProgress {
+		jobRegistry := c.destSysServer.JobRegistry().(*jobs.Registry)
+		job, err := jobRegistry.LoadJob(context.Background(), jobspb.JobID(ingestionJobID))
+		require.NoError(t, err)
+
+		progress := job.Progress()
+		ingestProgress := progress.Details.(*jobspb.Progress_StreamIngest).StreamIngest
+		return ingestProgress
+	}
+
+	// Since the ingestion job may retry at any moment, the producer PTS should never exceed
+	// the persisted checkpoint of the ingestion job
+	t.Run(`pts record never exceeds persisted ingestion checkpoint`, func(t *testing.T) {
+		args := defaultTenantStreamingClustersArgs
+		args.destInitFunc = func(t *testing.T, sysSQL *sqlutils.SQLRunner) {
+			settings := destClusterSetting
+			// Make heartbeats far more frequent than job checkpointing to ensure the
+			// producer's pts record moves as far forward as it can
+			settings[`stream_replication.job_checkpoint_frequency`] = `'1s'`
+			settings[`stream_replication.consumer_heartbeat_frequency`] = `'50ms'`
+			sysSQL.ExecMultiple(t, configureClusterSettings(settings)...)
+		}
+		c, cleanup := createTenantStreamingClusters(ctx, t, args)
+		producerJobID, ingestionJobID := c.startStreamReplication()
+
+		// Should take around 5s total
```

Comment:
Do you think it would make sense to have this logic be more of a unit test as opposed to heavy (and slower) integration test?
We want to make sure that persistedHighwater is updated iff maybeUpdateProgress actually updated progress.... Maybe we can test just that?

Reply:
Converted it to a test in the frontier_processor_test where its reusing a cluster across multiple tests 👍